### ISSUE TITLE
Ruby 2.3 support

### ISSUE THREAD

**Issue by voxik** [swig/swig#593]
Swig does not support Ruby 2.3 properly. I observe several test failures:

```
checking ruby testcase li_std_multimap (with run test)
li_std_multimap_wrap.cxx: In function 'VALUE std_multimap_Sl_int_Sc_A_Sm__Sg____getitem__(const std::multimap<int, A*, std::less<int>, std::allocator<std::pair<const int, A*> > >*, const key_type&)':
li_std_multimap_wrap.cxx:4500:14: error: '__builtin_choose_expr' was not declared in this scope
      return RARRAY_PTR(ary)[0];
              ^
../../../Examples/Makefile:1073: recipe for target 'ruby_cpp' failed
make[2]: *** [ruby_cpp] Error 1
Makefile:47: recipe for target 'li_std_multimap.cpptest' failed
make[1]: *** [li_std_multimap.cpptest] Error 2
```

or

```
checking ruby testcase std_containers (with run test)
std_containers_wrap.cxx: In function 'VALUE std_multimap_Sl_std_string_Sc_int_Sg____getitem__(const std::multimap<std::__cxx11::basic_string<char>, int, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, int> > >*, const key_type&)':
std_containers_wrap.cxx:5667:14: error: '__builtin_choose_expr' was not declared in this scope
      return RARRAY_PTR(ary)[0];
              ^
../../../Examples/Makefile:1073: recipe for target 'ruby_cpp' failed
make[2]: *** [ruby_cpp] Error 1
Makefile:47: recipe for target 'std_containers.cpptest' failed
make[1]: *** [std_containers.cpptest] Error 2
```

Searching for the cause, I found similar issue with explanation https://github.com/SciRuby/nmatrix/issues/429#issuecomment-169563925 and related PR https://github.com/SciRuby/nmatrix/pull/433 fixing the issue.

Swig probably needs similar treatment. Not sure about the backward compatibility though.

**Comment by mtasaka** [comment]
I think this is ruby issue:
https://bugs.ruby-lang.org/issues/11962
https://github.com/ruby/ruby/commit/98e565ec78cb4a07ffde8589ac4581fca31e9c17

**Comment by voxik** [comment]
@mtasaka you are right. Closing this. Sorry for the noise.

**Comment by ojwb** [comment]
While this is a Ruby issue, if there's a reasonably sane workaround it would be good to add it to SWIG so that people can use SWIG and Ruby 2.3.0 together.

**Comment by ojwb** [comment]
It looks to me like SWIG could include `<ruby.h>` like so:

```
#ifdef __cplusplus
// Work around C++ incompatibility in Ruby 2.3.0 headers:
// https://bugs.ruby-lang.org/issues/11962
# include "ruby/config.h"
# ifdef HAVE_BUILTIN___BUILTIN_CHOOSE_EXPR_CONSTANT_P
#  undef HAVE_BUILTIN___BUILTIN_CHOOSE_EXPR_CONSTANT_P
# endif
# ifdef HAVE_BUILTIN___BUILTIN_TYPES_COMPATIBLE_P
#  undef HAVE_BUILTIN___BUILTIN_TYPES_COMPATIBLE_P
# endif
#endif
#include <ruby.h>
```

I don't have Ruby 2.3.0 installed here currently, so I can't easily test that though - can someone affected try that?  You can just edit the SWIG generated C++ output by hand to do so.

**Comment by wsfulton** [comment]
Which version of Ruby is the fix in, is it later versions of 2.3.0 or is 2.3.1 required?

**Comment by mtasaka** [comment]
Looks like the fix for include/ruby/ruby.h ( https://github.com/ruby/ruby/commit/98e565ec78cb4a07ffde8589ac4581fca31e9c17 ) is not yet backported into branches/ruby_2_3/ .

At least ruby 2.3.0 has this bug, not sure if ruby 2.3.1 will fix this or not.

**Comment by voxik** [comment]
@wsfulton I requested here:

https://bugs.ruby-lang.org/issues/11962#note-5

**Comment by wsfulton** [comment]
I can't replicate this on my own machine nor on Travis (using the Ubuntu ppa:brightbox/ruby-ng ppa). I had to make some changes to configure.ac to detect ruby-2.3 though. I've added these changes into master. Are you sure your ruby build is installed correctly or maybe this is OS specific? Can you make sure you see the 2.3 headers and library being used when you are configuring. I get:
```
checking for Ruby header files... -I/usr/include/ruby-2.3.0 -I/usr/include/x86_64-linux-gnu/ruby-2.3.0
checking for Ruby library... -lruby-2.3 in /usr/lib/x86_64-linux-gnu
```
Likewise when you are running a test. 

See the new Travis ruby 2.3 build I added too - https://travis-ci.org/swig/swig/jobs/107173331

**Comment by ojwb** [comment]
I can reproduce on Debian unstable with current git master.  I did:

```
make distclean
./configure RUBY=/usr/bin/ruby2.3
make
cd Examples/test-suite/ruby
make li_std_multimap.cpptest
```

Output is:

```
if [ -f ./li_std_multimap_runme.rb ]; then echo "checking ruby testcase li_std_multimap (with run test)" ; else echo "checking ruby testcase li_std_multimap" ; fi;
checking ruby testcase li_std_multimap (with run test)
make -f ../../../Examples/Makefile SRCDIR='./' CXXSRCS='' SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' INCLUDES='-I../../../Examples/test-suite' SWIGOPT='-outcurrentdir -I../../../Examples/test-suite -w801 -noautorename -features autodoc=4' NOLINK=true TARGET='li_std_multimap' INTERFACEDIR='../' INTERFACE='li_std_multimap.i' ruby_cpp
make[1]: Entering directory '/home/olly/git/swig/Examples/test-suite/ruby'
env SWIG_LIB=../../../Lib  ../../../swig -c++ -ruby -outcurrentdir -I../../../Examples/test-suite -w801 -noautorename -features autodoc=4 -o li_std_multimap_wrap.cxx ./../li_std_multimap.i
g++ -c -fpic -I. -I./ -I/usr/include  -fPIC  li_std_multimap_wrap.cxx   -I../../../Examples/test-suite -I/usr/include/ruby-2.3.0 -I/usr/include/x86_64-linux-gnu/ruby-2.3.0
li_std_multimap_wrap.cxx: In function ‘VALUE std_multimap_Sl_int_Sc_A_Sm__Sg____getitem__(const std::multimap<int, A*, std::less<int>, std::allocator<std::pair<const int, A*> > >*, const key_type&)’:
li_std_multimap_wrap.cxx:4611:166: error: ‘__builtin_choose_expr’ was not declared in this scope
../../../Examples/Makefile:1073: recipe for target 'ruby_cpp' failed
make[1]: *** [ruby_cpp] Error 1
make[1]: Leaving directory '/home/olly/git/swig/Examples/test-suite/ruby'
Makefile:47: recipe for target 'li_std_multimap.cpptest' failed
make: *** [li_std_multimap.cpptest] Error 2
```

I'm getting the right headers, etc:

```
checking for ruby... /usr/bin/ruby2.3
checking for Ruby header files... -I/usr/include/ruby-2.3.0 -I/usr/include/x86_64-linux-gnu/ruby-2.3.0
checking for Ruby library... -lruby-2.3 in /usr/lib/x86_64-linux-gnu
```

**Comment by voxik** [comment]
This was fixed by Ruby 2.3.1, so this could be closed from my POV.

**Comment by ojwb** [comment]
Keeping this issue open doesn't really seem to serve a purpose - we know this is fixed by Ruby 2.3.1, and we have a proposed workaround for 2.3.0 - either we should explicitly decide that we don't support C++ bindings with Ruby 2.3.0 and GCC (which is probably a small intersection now 2.3.1 has been out for a while), or apply the workaround I suggested above.

@wsfulton While Ubuntu 16.04 has Ruby 2.3.0-based packages, they include the fix for this - that's probably why you couldn't reproduce this.

**Comment by ojwb** [comment]
Nobody spoke up for working around this, so just closing.
